File: paper/benchmarks/bench_chained.py
"""Benchmark AIP chained token creation, delegation, and verification (Python)."""
import sys, os, time, json, statistics
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'python'))

ITERATIONS = 100  # fewer iterations since chained ops are heavier

outdir = os.path.join(os.path.dirname(__file__), 'results')
os.makedirs(outdir, exist_ok=True)

try:
    from aip_core.crypto import KeyPair
    from aip_token.chained import ChainedToken
except ImportError:
    print("WARNING: biscuit-python not installed, skipping chained benchmark.")
    results = {"language": "python", "mode": "chained", "skipped": True,
               "reason": "biscuit-python not installed", "depths": []}
    print(json.dumps(results, indent=2))
    with open(os.path.join(outdir, 'chained_python.json'), 'w') as f:
        json.dump(results, f, indent=2)
    sys.exit(0)

MAX_DEPTH = 5

depths = []

for depth in range(MAX_DEPTH + 1):
    create_times = []
    verify_times = []
    token_sizes = []

    for _ in range(ITERATIONS):
        kp = KeyPair.generate()
        pubkey = kp.public_key_bytes()

        # Create authority token (depth 0)
        start = time.perf_counter_ns()
        token = ChainedToken.create_authority(
            issuer="aip:web:bench.test/agent-0",
            scopes=["tool:search", "tool:browse"],
            budget_cents=1000,
            max_depth=MAX_DEPTH,
            ttl_seconds=3600,
            keypair=kp,
        )
        authority_create_ns = time.perf_counter_ns() - start

        # Delegate up to the target depth
        for d in range(1, depth + 1):
            start = time.perf_counter_ns()
            token = token.delegate(
                delegator=f"aip:web:bench.test/agent-{d-1}",
                delegate=f"aip:web:bench.test/agent-{d}",
                scopes=["tool:search"],
                budget_cents=1000,
                context=f"delegation at depth {d}",
            )
            append_ns = time.perf_counter_ns() - start

        if depth == 0:
            create_times.append(authority_create_ns / 1_000_000)
        else:
            create_times.append(append_ns / 1_000_000)

        # Measure token size
        b64 = token.to_base64()
        token_sizes.append(len(b64.encode('utf-8')))

        # Measure verification: from_base64 + authorize
        start = time.perf_counter_ns()
        restored = ChainedToken.from_base64(b64, pubkey)
        restored.authorize("tool:search", pubkey)
        verify_times.append((time.perf_counter_ns() - start) / 1_000_000)

    entry = {
        "depth": depth,
        "token_size_bytes": round(statistics.mean(token_sizes)),
    }

    if depth == 0:
        entry["create_ms"] = round(statistics.mean(create_times), 4)
    else:
        entry["append_ms"] = round(statistics.mean(create_times), 4)

    entry["verify_ms"] = round(statistics.mean(verify_times), 4)

    depths.append(entry)

results = {
    "language": "python",
    "mode": "chained",
    "iterations_per_depth": ITERATIONS,
    "depths": depths,
}
print(json.dumps(results, indent=2))

with open(os.path.join(outdir, 'chained_python.json'), 'w') as f:
    json.dump(results, f, indent=2)

File: rust/aip-token/src/bin/bench_chained.rs
//! Benchmark AIP chained token creation, delegation, and verification (Rust).

use aip_core::crypto::KeyPair;
use aip_token::chained::ChainedToken;
use std::time::Instant;

const MAX_DEPTH: u32 = 5;
const ITERATIONS: usize = 100;

fn main() {
    println!("{{");
    println!("  \"language\": \"rust\",");
    println!("  \"mode\": \"chained\",");
    println!("  \"iterations_per_depth\": {},", ITERATIONS);
    println!("  \"depths\": [");

    for depth in 0..=MAX_DEPTH {
        let mut create_times = Vec::new();
        let mut verify_times = Vec::new();
        let mut token_sizes = Vec::new();

        for _ in 0..ITERATIONS {
            let kp = KeyPair::generate();
            let pubkey = kp.public_key_bytes();

            // Create authority token (depth 0)
            let start = Instant::now();
            let mut token = ChainedToken::create_authority(
                "aip:web:bench.test/agent-0",
                &["tool:search", "tool:browse"],
                Some(1000),
                MAX_DEPTH,
                3600,
                &kp,
            )
            .unwrap();
            let authority_elapsed = start.elapsed().as_nanos() as f64 / 1_000_000.0;

            // Delegate up to the target depth
            let mut append_elapsed = 0.0;
            for d in 1..=depth {
                let delegator = format!("aip:web:bench.test/agent-{}", d - 1);
                let delegate = format!("aip:web:bench.test/agent-{}", d);
                let context = format!("delegation at depth {}", d);

                let start = Instant::now();
                token = token
                    .delegate(
                        &delegator,
                        &delegate,
                        &["tool:search"],
                        Some(1000),
                        &context,
                    )
                    .unwrap();
                append_elapsed = start.elapsed().as_nanos() as f64 / 1_000_000.0;
            }

            if depth == 0 {
                create_times.push(authority_elapsed);
            } else {
                create_times.push(append_elapsed);
            }

            // Measure token size (base64 length)
            let b64 = token.to_base64().unwrap();
            token_sizes.push(b64.len());

            // Measure verification: from_base64 + authorize
            let start = Instant::now();
            let restored = ChainedToken::from_base64(&b64, &pubkey).unwrap();
            restored.authorize("tool:search", &pubkey).unwrap();
            verify_times.push(start.elapsed().as_nanos() as f64 / 1_000_000.0);
        }

        let avg_create: f64 = create_times.iter().sum::<f64>() / ITERATIONS as f64;
        let avg_verify: f64 = verify_times.iter().sum::<f64>() / ITERATIONS as f64;
        let avg_size: usize = token_sizes.iter().sum::<usize>() / ITERATIONS;

        let comma = if depth < MAX_DEPTH { "," } else { "" };

        if depth == 0 {
            println!(
                "    {{\"depth\": {}, \"token_size_bytes\": {}, \"create_ms\": {:.4}, \"verify_ms\": {:.4}}}{}",
                depth, avg_size, avg_create, avg_verify, comma
            );
        } else {
            println!(
                "    {{\"depth\": {}, \"token_size_bytes\": {}, \"append_ms\": {:.4}, \"verify_ms\": {:.4}}}{}",
                depth, avg_size, avg_create, avg_verify, comma
            );
        }
    }

    println!("  ]");
    println!("}}");
}
